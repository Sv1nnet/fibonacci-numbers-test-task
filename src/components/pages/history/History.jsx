import React, { useState, useEffect } from 'react';
import RedirectButton from '@components/ui/buttons/redirectButton/RedirectButton';
import HistoryItem from '@components/pages/history/historyItem/HistoryItem';
import { Pagination } from 'react-materialize';
import Icon from '@components/ui/icons/arrow/Icon';
import { useGetHistory } from '@/hooks/api/apiHooks';

import './history.scss';

const History = () => {
  const [pageNumber, setPageNumber] = useState(1);
  const [{
    data = {
      results: [],
      maxPages: 1,
      page: 1,
    },
    loading,
    error,
  }, getHistory] = useGetHistory();

  useEffect(() => {
    getHistory(pageNumber);
  }, [pageNumber, getHistory]);

  return (
    <div className="history">
      {
        data.results.length > 0
          ? (
            <>
              <table>
                <thead>
                  <tr>
                    <th>Number</th>
                    <th className="history-list-item--result-header">Results</th>
                  </tr>
                </thead>

                <tbody>
                  {
                    data.results.map((res) => (
                      <HistoryItem key={res.id} number={res.number} result={res.result} />
                    ))
                  }
                </tbody>
              </table>
              <Pagination
                className="history-pag"
                activePage={pageNumber}
                items={data.maxPages}
                leftBtn={<Icon />}
                maxButtons={8}
                rightBtn={<Icon className="arrow-right" />}
                onSelect={setPageNumber}
              />
              {loading && <p className="history-loading">Loading...</p>}
            </>
          )
          : (
            <p className={`history-loading ${error ? 'red-text accent-4' : ''}`}>
              {
                (function defineHistoryLoadingContent() {
                  let historyLoadingContent = 'Loading...';

                  if (error) historyLoadingContent = error.message || error.statusText || 'Unexpected Error';
                  else if (!loading) historyLoadingContent = 'You have no calculations yet';

                  return historyLoadingContent;
                }())
              }
            </p>
          )
      }
      <RedirectButton className="history-redir-btn" to="/" text="Back to main" />
    </div>
  );
};

export default History;
