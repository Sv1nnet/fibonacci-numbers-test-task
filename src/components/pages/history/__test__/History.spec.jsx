import {
  it,
  describe,
  expect,
} from '@jest/globals';
import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import History from '../History';

configure({ adapter: new Adapter() });

describe('Test History component', () => {
  it('Should match snapshot with', () => {
    const component = shallow(<History />, {
      wrappingComponent: MemoryRouter,
    });
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Should download history and update state. DB has to have at least one record for the IP which we make request from, and server has to be running', async () => {
    const { container } = render(
      <MemoryRouter>
        <History />
      </MemoryRouter>,
    );
    await screen.findByText('Results');
    expect(container.querySelector('table')).toBeTruthy();
  });
});
