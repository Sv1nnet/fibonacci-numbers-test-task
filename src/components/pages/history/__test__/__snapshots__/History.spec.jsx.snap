// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test History component Should match snapshot with 1`] = `
<div
  className="history"
>
  <p
    className="history-loading "
  >
    You have no calculations yet
  </p>
  <RedirectButton
    className="history-redir-btn"
    text="Back to main"
    to="/"
  />
</div>
`;
